### ISSUE TITLE
Suspense!

### ISSUE THREAD

**Issue by jaredpalmer** [jaredpalmer/after.js#175]
Opening up discussion for what Suspense-based API will look like.

**Comment by krazyjakee** [comment]
@jaredpalmer at a high level, how do you envisage this affecting afterjs? What could be the benefits?

**Comment by jaredpalmer** [comment]
single pass rendering.

**Comment by jaredpalmer** [comment]
no need for after.js lol

**Comment by jaredpalmer** [comment]
but seriously, there will still need to be some way to hydrate even with streaming. so after.js would handle that if React pushes some of that to user land. it's hard to say.

**Comment by krazyjakee** [comment]
"fizz" updates have been made to enable a better path for sssr https://github.com/facebook/react/pull/14144

**Comment by renatorroliveira** [comment]
Man, so many things happening, hard to follow hahaha

So, I am just researching a better way to migrate an application on React 16 + React Router 4 + Redux Sagas to a SSR solution. I was almost decided to go with After since Next.js is "too opinionated" for my taste and Electrode seems "too complicated" (I think of it like a war machine for my simples SSR for SEO needs... maybe I am mistaken, but the bootstrapped app seems already too complicated).

Should I wait or try using alpha versions of React 17? I am divided right know between After and try React 17 by hand. @jaredpalmer It would be too hard to migrate an After based SSR app to the new approach introduced by React 17?

**Comment by krazyjakee** [comment]
@renatorroliveira check my comment here https://github.com/jaredpalmer/after.js/issues/133#issuecomment-450164506

As I see it essentially removes the need for getInitialProps which both next and Afterjs use.

I'm very sure it will be simple to move logic out of getInitialProps and into a function called by suspense. Both return props so I'm confident it would be simple.

**Comment by jaredpalmer** [comment]
It's hard to say for sure, but yes we could do what @krazyjakee described. However, the entire premise of `getInitialProps` would be theoretically void when SSR suspense arrives because you no longer would be forced to inject data top-down.

**Comment by krazyjakee** [comment]
@jaredpalmer unrelated question. In a post-getInitialProps world, where would a component receive the [afterjs context](https://github.com/jaredpalmer/after.js/blob/master/examples/basic/src/server.js#L18-L20) to, for example, set the response status code?

**Comment by krazyjakee** [comment]
Nooooo don't you dare. Palmdawg needs this issue.

**Comment by Ricki-BumbleDev** [comment]
Just found [react-lightyear](https://www.npmjs.com/package/react-lightyear), which exposes a `renderToStringAsync` function and threw it together with [fetch-suspense](https://www.npmjs.com/package/fetch-suspense), which exposes a `useFetch` hook supporting Suspense. And it actually works! (Only thing I had to do is to tweak fetch-suspense to use [isomorphic-fetch](https://www.npmjs.com/package/isomorphic-fetch))

**Comment by krazyjakee** [comment]
@Ricki-BumbleDev got a minimal project we can look at?

**Comment by Ephem** [comment]
@Ricki-BumbleDev Glad you found it useful! 😄

@krazyjakee There are also a couple of examples in the repo itself, for using with [Redux](https://github.com/Ephem/react-lightyear/tree/lightyear/examples/redux), [Apollo (using the unofficial react-apollo-hooks)](https://github.com/Ephem/react-lightyear/tree/lightyear/examples/react-apollo-hooks) or [urql](https://github.com/Ephem/react-lightyear/tree/lightyear/examples/urql).

Checking out After.js has been on my list for a while but I haven't gotten around to it. After finding this issue I skimmed the Custom/Async Rendering-part of the documentation and there should be no problem just switching out the `renderToString` from React to `renderToStringAsync` from `react-lightyear` instead and everything should just work!
